lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -192,37 +191,45 @@ internal class EvaluatingCompiler(
             }
         }
 
-        override fun compute() = current?.exprValue() ?: valueFactory.nullValue
+        override fun compute() = current ?: valueFactory.nullValue
     }
 
-    private fun comparisonAccumulator(cmpFunc: (Number, Number) -> Boolean): (Number?, ExprValue) -> Number =
-        { curr, next ->
-            val nextNum = next.numberValue()
-            when (curr) {
-                null -> nextNum
-                else -> when {
-                    cmpFunc(nextNum, curr) -> nextNum
-                    else -> curr
-                }
+    private fun comparisonAccumulator(comparator: NaturalExprValueComparators): (ExprValue?, ExprValue) -> ExprValue =
+        { left, right ->
+            when {
+                left == null || comparator.compare(left, right) > 0 -> right
+                else -> left
             }
         }
 
     /** Dispatch table for built-in aggregate functions. */
     private val builtinAggregates: Map<Pair<String, PartiqlAst.SetQuantifier>, ExprAggregatorFactory> =
         run {
-            val countAccFunc: (Number?, ExprValue) -> Number = { curr, _ -> curr!! + 1L }
-            val sumAccFunc: (Number?, ExprValue) -> Number = { curr, next ->
-                curr?.let { it + next.numberValue() } ?: next.numberValue()
+            fun checkIsNumberType(funcName: String, value: ExprValue) {
+                if (!value.type.isNumber) {
+                    throw errNoContext(
+                        message = "Aggregate function $funcName expects arguments of NUMBER type but the following value was provided: ${value.ionValue}, with type of ${value.type}",
+                        errorCode = ErrorCode.EVALUATOR_UNEXPECTED_VALUE_TYPE,
```

Comment:
I wonder if this is the appropriate error code?  Remember that error codes also have their own messages associated with them, with the intent of making them localizable.  Some customers use the exception message but there's at least one that I know of that uses the error messages associated with the error code.  Anyway, the error message associated with this code is: "Unexpected value type".  Consider adding a new error code with a similar / friendly error message.